File: 이준영/백준/Platinum/5670. 휴대폰 자판/휴대폰 자판.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.StringTokenizer;

public class Main {
	static int count;

	public static class Node {
		HashMap<Character, Node> child;
		boolean end;
		int child_count;

		public Node() {
			this.child = new HashMap<>();
			this.end = false;
			this.child_count = 0;
		}
	}

	public static class Trie {
		Node root;

		public Trie() {
			this.root = new Node();
		}

		public void insert(String str) {
			Node node = this.root;
			for (int i = 0; i < str.length(); i++) {
				node.child_count += 1;
				char c = str.charAt(i);
				node.child.putIfAbsent(c, new Node());
				node = node.child.get(c);
			}
			node.child_count+=1;
			node.end = true;
		}

		public void find_sum(Node node) {
			if (node.child.isEmpty()) {
				return;
			}
			if (node.child.size() >= 2 || node == root || node.end) {
				for (Node tmp : node.child.values()) {
					count += tmp.child_count;
					find_sum(tmp);
				}
			} else {
				for (Node tmp : node.child.values()) {
					find_sum(tmp);
				}
			}
			return;
		}
	}

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		StringBuilder sb1 = new StringBuilder();
		while (true) {
			try {
				count = 0;
				st = new StringTokenizer(br.readLine());
				int N = Integer.parseInt(st.nextToken());
				Trie trie = new Trie();
				
				for (int i = 0; i < N; i++) {
					st = new StringTokenizer(br.readLine());
					String str = st.nextToken();
					trie.insert(str);
				}
				
				trie.find_sum(trie.root);
				
				double result = (double) count / N * 100;
				result = (double) Math.round(result) / 100;
				String formattedNumber = String.format("%.2f", result);
				sb1.append(formattedNumber).append("\n");
			} catch (Exception e) {
//				System.out.println(e);
				break;
			}
		}
		System.out.println(sb1);
	}

}

File: 이준영/백준/Platinum/5670. 휴대폰 자판/휴대폰 자판.py
import sys

class Node:
    def __init__(self):
        self.child = {}        # 자식 노드 맵
        self.end = False       # 단어 종료 표시
        self.child_count = 0   # 이 노드를 거친 단어 수

class Trie:
    def __init__(self):
        self.root = Node()
        self.count = 0

    def insert(self, s: str):
        node = self.root
        for c in s:
            node.child_count += 1
            node.child.setdefault(c, Node())
            node = node.child[c]
        node.child_count += 1
        node.end = True

    def find_sum(self):
        def dfs(node: Node):
            if not node.child:
                return
            if node is self.root or node.end or len(node.child) >= 2:
                for nxt in node.child.values():
                    self.count += nxt.child_count
                    dfs(nxt)
            else:
                for nxt in node.child.values():
                    dfs(nxt)
        dfs(self.root)

def main():
    rd = sys.stdin.readline
    out_lines = []

    while True:
        try:
            line = rd()
            # Java의 br.readLine()에서 null 반환시 catch로 빠지지만,
            # Python은 빈 문자열 반환 → EOF 처리
            if not line:
                break

            N = int(line.strip())
            trie = Trie()

            for _ in range(N):
                word = rd().strip()
                trie.insert(word)

            trie.find_sum()
            avg = trie.count / N
            out_lines.append(f"{avg:.2f}")

        except Exception:
            # 입력 포맷 오류나 EOF 발생 시 루프 종료
            break

    sys.stdout.write("\n".join(out_lines))

if __name__ == "__main__":
    main()
